### PR TITLE
Add apiserver heartbeat with conntrack

### LINKED ISSUE #9792: Unable update CRD identity information with a reference for this node
User report from Slack:

> Curious if anyone has seen an error like `Unable update CRD identity information with a reference for this node` in their cilium pod logs? Might be a red herring. I'm running Cilium on EKS in CNI chaining mode. It's been fine for a couple of weeks, but today it degraded to a state where if a pod gets scheduled on a node, it gets stuck in ContainerCreating. In the pod's events, I see errors like:
> `NetworkPlugin cni failed to set up pod "REDACTED" network: unable to create endpoint: Put http:///var/run/cilium/cilium.sock/v1/endpoint/cilium-local:0: context deadline exceeded`.
> If I delete the cilium pod on the node, all other stuck pods get scheduled successfully once the cilium pod comes up again.

> I see a lot of this error as well
> `2019-12-04T23:00:19.707077314Z level=warning msg="Key allocation attempt failed" attempt=11 error="unable to create slave key 'k8s:app=REDACTED;k8s:io.cilium.k8s.policy.cluster=default;k8s:io.cilium.k8s.policy.serviceaccount=REDACTED;k8s:io.kubernetes.pod.namespace=default;k8s:service-type=rest;': Operation cannot be fulfilled on ciliumidentities.cilium.io \"64380\": StorageError: invalid object, Code: 4, Key: /registry/cilium.io/ciliumidentities/64380, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: f6d81d4c-0a28-11ea-b6ee-122bfb3499ba, UID in object meta: " key="[k8s:app=REDACTED k8s:io.cilium.k8s.policy.cluster=default k8s:io.cilium.k8s.policy.serviceaccount=REDACTED k8s:io.kubernetes.pod.namespace=default k8s:service-type=rest]" subsys=allocator`
> The error message `StorageError: invalid object, Code: 4` seems to come from kubernetes/kube-apiserver. I also notice that the crd kvstore implementation uses a cache in front of kube-api. I’m wondering if there could be some cache invalidation issues at play here.

> So we think we know what causes this issue, Amazon periodically upgrades the “eks-version” which includes upgrading the control plane to a newer patch version.
> During this time the control plane is restarted by AWS
> These changes are not rolled out to all EKS clusters at the same, they apparently do some kind of a random rollout process. Today two of our clusters got upgraded from eks.2->eks.6. After some pod churn both clusters starting seeing pods being failed to start.
> We also believe a similar upgrade happened around the time when we first saw this issue a few weeks ago (we think that the version went from 1->2).

### PR DESCRIPTION
**Pull request by tom-hadlaw-hs**
:warning: :warning: **aanm note to reviewers / backporters**: This fixes a bug that needs to be backported to 1.7/1.6 but we should only do it after heavily testing it.

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](http://docs.cilium.io/en/stable/contributing/contributing/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue. 
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](http://docs.cilium.io/en/stable/contributing/contributing/#dev-coo).
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

Adding a periodic heartbeat to check for stale connections to the apiserver. If a request takes too long then client-go's conntrack dialer close connections function is called for all client types.

Fixes: #9792

```release-note
When running in Kubernetes, Cilium will run a periodic heartbeat and close all open Kubernetes client connections if the active connections become unresponsive. 
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/10184)

### CODE REVIEW

**Comment by tom-hadlaw-hs** [comment]
@aanm I think it might make sense to use context.WithTimeout to achieve a heartbeat timeout. I notice that client-go has added supporting context.Context very recently upstream. When does the forked client-go get changes merged in, is this something I could do?

**Review by aanm — COMMENTED** [review]
This is a move in the right direction! Maybe we should add a cilium-agent flag that allows to user to specify the timeout and keep alive. Just to be sure, are the timeouts the same used by kubelet?

**Comment by aanm** [comment]
> @aanm I think it might make sense to use context.WithTimeout to achieve a heartbeat timeout. I notice that client-go has added supporting context.Context very recently upstream. When does the forked client-go get changes merged in, is this something I could do?

The `context.WithTimeout` could fail for other reasons, e.g. the server was overloaded to serve the request, in this case it would be better to not try to create a new connection or we will overload the server even more.

Adding support for `context.Context` in Cilium so it can be used with kubernetes library is very welcome.

**Comment by tom-hadlaw-hs** [comment]
@aanm Yeah, it's a bit tricky as a network connection silently failing causes calls to apiserver to hang for as long as the kernel doesn't time out the connection. I'm not sure if using context.WithTimeout would solve this issue but i'll play around with it. I'm wondering if that's just a side effect of client-go's retry logic built on top of golangs net libraries. 

Adding a flag makes sense, I'll add that.

Kubelet seems to just use a basic client so I'm actually not sure if/how it would avoid the same situation we're seeing. I'll look into it deeper to find out. Currently it just uses `context.TODO()`.

**Comment by tom-hadlaw-hs** [comment]
For reference, heres where the Kubelet performs its heartbeat: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L423

What are your thoughts about a flag that enables/disables the heartbeat feature?

**Comment by tom-hadlaw-hs** [comment]
I decided to use the underlying client-go rest client to just request the healthz endpoint, instead of making actual resource requests, as I'm worried this could be adding too much load on the apiserver with a sufficiently sized cluster.
 
It deviates with kubelet's implementation of this but as far as I can tell their approach doesn't actually solve the issue described in that thread.

File: daemon/daemon_main.go [hunk 1]
```diff
@@ -712,6 +712,11 @@ func init() {
 	option.BindEnv(option.PolicyTriggerInterval)
 
 	flags.Bool(option.DisableCNPStatusUpdates, false, "Do not send CNP NodeStatus updates to the Kubernetes api-server (recommended to run with `cnp-node-status-gc=false` in cilium-operator)")
+
+	flags.Bool(option.EnableK8sHeartbeat, true, "Run periodic heartbeat against apiserver which resets all open connection after timeout")
+
+	flags.Duration(option.K8sHeartbeatTimeout, 30*time.Second, "Timeout for k8s heartbeat")
+
```
Comment (aanm) [inline daemon/daemon_main.go]: please put this after line 720 and add a similar line as `option.BindEnv...` for each option

File: pkg/k8s/client_test.go [hunk 2]
```diff
@@ -29,7 +29,7 @@ import (
 	"github.com/cilium/cilium/pkg/source"
 
 	. "gopkg.in/check.v1"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (aanm) [inline pkg/k8s/client_test.go]: Unnecessary changes

File: pkg/k8s/init.go [hunk 3]
```diff
@@ -11,7 +11,6 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
-
```
Comment (aanm) [inline pkg/k8s/init.go]: please re-add this new line

File: daemon/daemon_main.go [hunk 4]
```diff
@@ -712,6 +712,11 @@ func init() {
 	option.BindEnv(option.PolicyTriggerInterval)
 
 	flags.Bool(option.DisableCNPStatusUpdates, false, "Do not send CNP NodeStatus updates to the Kubernetes api-server (recommended to run with `cnp-node-status-gc=false` in cilium-operator)")
+
+	flags.Bool(option.EnableK8sHeartbeat, true, "Run periodic heartbeat against apiserver which resets all open connection after timeout")
+
+	flags.Duration(option.K8sHeartbeatTimeout, 30*time.Second, "Timeout for k8s heartbeat")
```
Comment (aanm) [inline daemon/daemon_main.go]: We can drop `option.EnableK8sHeartbeat` since the user can disable this feature by setting `K8sHeartbeatTimeout` with 0.

File: pkg/k8s/client.go [hunk 5]
```diff
@@ -105,11 +109,66 @@ func CreateConfig() (*rest.Config, error) {
...
+	go wait.Until(func() {
+		done := make(chan error)
+		go func() {
+			err := restClient.Get().AbsPath("/healthz").Do().Error()
+			if err != nil {
+				done <- err
+			}
+
+			close(done)
+		}()
```
Comment (aanm) [inline pkg/k8s/client.go]: I'm afraid of this. It seems harmless but that won't be the case for big clusters. `/healthz` could return a 503 and we would restart the connection, possibly make it worse.

**Review by aanm — CHANGES_REQUESTED** [review]
@tom-hadlaw-hs this is on the right path. I'm just a little concerned with the `runHeartbeat` function. Can you try it out in your cluster without the `runHeartbeat` to see if the issue still persists? I can provision a cilium image if you make a commit reverting the `runHeartbeat` for you to try out.

**Comment by tom-hadlaw-hs** [comment]
@aanm I just re-ran the upgrade process test, I can no longer reproduce this issue using this method. It seems as though Amazon might've changed the behavior of their rollout process. 

After the upgrade completes, cilium-agent pods restart with the following message:

`error: http2: server sent GOAWAY and closed the connection; LastStreamID=3, ErrCode=NO_ERROR, debug=""`

So it looks like Amazon is trying to prevent similar issues from happening by notifying that the connection is about to close.

Note: I've been testing my changes by adding IP table rules blocking traffic to kubernetes service endpoints, which functionally appears to do the same thing as EKS.

File: pkg/k8s/client.go (hunk 5 above)
Comment (tgraf) [reply pkg/k8s/client.go]: How does kubelet cope with this?

**Comment by tom-hadlaw-hs** [comment]
I also wanted to mention that I've found an issue/PR on golang net library itself which seeks add application layer pinging on http2 connections to fix issues such as this one. This would likely be the ideal solution to all this, whenever that change is released.

File: pkg/k8s/client.go (hunk 5 above)
Comment (tom-hadlaw-hs) [reply pkg/k8s/client.go]: Kubelet does a get on nodes

File: daemon/daemon_main.go [hunk 6]
```diff
@@ -714,6 +714,9 @@ func init() {
 	flags.Bool(option.DisableCNPStatusUpdates, false, "Do not send CNP NodeStatus updates to the Kubernetes api-server (recommended to run with `cnp-node-status-gc=false` in cilium-operator)")
 	option.BindEnv(option.DisableCNPStatusUpdates)
 
+	flags.Duration(option.K8sHeartbeatTimeout, 30*time.Second, "Timeout for k8s heartbeat")
+	option.BindEnv(option.K8sHeartbeatTimeout)
+
```
Comment (aanm) [inline daemon/daemon_main.go]: Please also add a flag in `operator/flags.go`

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/k8s/client.go [hunk 7]
```diff
@@ -105,11 +108,50 @@ func CreateConfig() (*rest.Config, error) {
...
+	config.Dial = dialer.DialContext
+	return dialer.CloseAll
+}
+
+func runHeartbeat(closeAllConns []func(), stop chan struct{}) error {
+	timeout := option.Config.K8sHeartbeatTimeout
+	go wait.Until(func() {
+		done := make(chan error)
+		go func() {
+			_, err := k8sCli.Interface.CoreV1().Namespaces().Get("kube-system", metav1.GetOptions{})
```
Comment (aanm) [inline pkg/k8s/client.go:133]: please add this comment to the code:
```
Kubernetes does a get node of the node that kubelet is running [0]. This seems excessive in our case because the amount of data transferred is bigger than doing a Get of the kube-system namespace. For this reason we have picked to perform a get on `kube-system` instead a get of a node.

[0] https://github.com/kubernetes/kubernetes/blob/v1.17.3/pkg/kubelet/kubelet_node_status.go#L423
```

File: pkg/k8s/client.go [hunk 8]
```diff
@@ -105,11 +108,50 @@ func CreateConfig() (*rest.Config, error) {
 	return createConfig(GetAPIServerURL(), GetKubeconfigPath(), GetQPS(), GetBurst())
 }
 
+func setDialer(config *rest.Config) func() {
+	context := (&net.Dialer{
+		Timeout:   30 * time.Second,
```
Comment (aanm) [inline pkg/k8s/client.go]: Shouldn't this be `Timeout: option.Config.K8sHeartbeatTimeout`?

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/k8s/client.go [hunk 9]
```diff
@@ -105,11 +108,50 @@ func CreateConfig() (*rest.Config, error) {
...
+			if err != nil {
+				done <- err
+			}
+
+			close(done)
+		}()
+
+		select {
+		case err := <-done:
+			if err != nil {
```
Comment (aanm) [inline pkg/k8s/client.go:147]: In case the error is type of timeout (504 and / or 408?), we won't execute `closeAllConns`, is this expected?

File: pkg/k8s/client.go (hunk 8 above)
Comment (tom-hadlaw-hs) [reply pkg/k8s/client.go]: This is just for dial timeouts, 30s was the kubelet hardcoded default. I don't think we want to set this the same as the heartbeat timeout as that would affect all client connections, especially in the case where you might set your timeout to be zero.

Comment (aanm) [reply pkg/k8s/client.go]: Oh right!

File: daemon/daemon_main.go (hunk 6 above)
Comment (tom-hadlaw-hs) [reply daemon/daemon_main.go]: done

File: pkg/k8s/client.go (hunk 9 above)
Comment (tom-hadlaw-hs) [reply pkg/k8s/client.go:147]: Yeah that makes sense, I changed it such that if there is a status error and either one of those status-codes, it'll restart the connection as well. Perhaps we'd want to add `StatusBadGateway` to that list?

Comment (aanm) [reply pkg/k8s/client.go:147]: Yeah I guess we can add that one as well.

**Comment by aanm** [comment]
test-me-please

**Comment by tom-hadlaw-hs** [comment]
Rebased the branch, seems like the Ginko tests where failing to provision test servers when it last failed.

**Comment by aanm** [comment]
test-me-please

**Comment by tom-hadlaw-hs** [comment]
Tests where failing on a documentation diff test, should be solved now @aanm

**Comment by joestringer** [comment]
test-me-please

**Comment by aanm** [comment]
```diff
01:12:13      runtime: ../operator/cilium-operator --cmdref /tmp/tmp.lIHAuZ3tWZ
01:12:13      runtime: make[1]: Leaving directory '/home/vagrant/go/src/github.com/cilium/cilium/Documentation'
01:12:13      runtime: diff -x '*.rst' -r ./Documentation/cmdref/cilium-agent.md /tmp/tmp.lIHAuZ3tWZ/cilium-agent.md
01:12:13      runtime: 108c108
01:12:13      runtime: <       --k8s-heartbeat-timeout                         Configures the timeout for api-server heartbeat, set to 0 to disable (default 30s)
01:12:13      runtime: ---
01:12:13      runtime: >       --k8s-heartbeat-timeout duration                Timeout for k8s heartbeat (default 30s)
01:12:13      runtime: diff -x '*.rst' -r ./Documentation/cmdref/cilium-operator.md /tmp/tmp.lIHAuZ3tWZ/cilium-operator.md
01:12:13      runtime: 41c41
01:12:13      runtime: <       --k8s-heartbeat-timeout                   Configures the timeout for api-server heartbeat, set to 0 to disable (default 30s)
01:12:13      runtime: ---
01:12:13      runtime: >       --k8s-heartbeat-timeout duration          Timeout for k8s heartbeat (default 30s)
01:12:13      runtime: Detected a difference in the cmdref directory
01:12:13      runtime: diff -r: diff -r ./Documentation/cmdref/cilium-agent.md /tmp/tmp.lIHAuZ3tWZ/cilium-agent.md
01:12:13      runtime: 108c108
01:12:13      runtime: <       --k8s-heartbeat-timeout                         Configures the timeout for api-server heartbeat, set to 0 to disable (default 30s)
01:12:13      runtime: ---
01:12:13      runtime: >       --k8s-heartbeat-timeout duration                Timeout for k8s heartbeat (default 30s)
01:12:13      runtime: diff -r ./Documentation/cmdref/cilium-operator.md /tmp/tmp.lIHAuZ3tWZ/cilium-operator.md
01:12:13      runtime: 41c41
01:12:13      runtime: <       --k8s-heartbeat-timeout                   Configures the timeout for api-server heartbeat, set to 0 to disable (default 30s)
01:12:13      runtime: ---
01:12:13      runtime: >       --k8s-heartbeat-timeout duration          Timeout for k8s heartbeat (default 30s)
01:12:13      runtime: Only in ./Documentation/cmdref: cli_index.rst
01:12:13      runtime: Only in ./Documentation/cmdref: index.rst
01:12:13      runtime: Only in ./Documentation/cmdref: kvstore.rst
```

@tom-hadlaw-hs this still requires documentation updates.

**Comment by tom-hadlaw-hs** [comment]
I hope that fixes it, I'm having trouble running the postcheck tests locally to verify

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Review by tgraf — APPROVED** [review]
